flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/strategy/SchedulingStrategyUtils.java

```diff
@@ -60,4 +61,18 @@
 				deploymentOptionRetriever.apply(executionVertexID)))
 			.collect(Collectors.toList());
 	}
+
+	static List<SchedulingPipelinedRegion<?, ?>> sortPipelinedRegionsInTopologicalOrder(
+			final SchedulingTopology<?, ?> topology,
+			final Set<SchedulingPipelinedRegion<?, ?>> regions) {
+
+		final Set<SchedulingPipelinedRegion<?, ?>> deduplicator = new HashSet<>();
+		return IterableUtils.toStream(topology.getVertices())
+			.map(SchedulingExecutionVertex::getId)
+			.map(topology::getPipelinedRegionOfVertex)
+			.filter(regions::contains)
+			.filter(region -> !deduplicator.contains(region))
+			.filter(deduplicator::add)
```

Comment:
Is the first line needed? `Set#add()` returns `false` if the item was not added.

Also, consider using `.distinct()` since side effects in stream operations are frowned upon:
> predicate – a non-interfering, stateless predicate to apply to each element to determine if it should be included
![image](https://user-images.githubusercontent.com/1681921/79491383-ae4ede80-801e-11ea-8c94-b07252003636.png)

Reply:
Good idea!